janus_server/src/client.rs

```diff
@@ -29,63 +37,215 @@ static CLIENT_USER_AGENT: &str = concat!(
+        label: Label::InputShare,
+        sender_role: Role::Client,
+        recipient_role: Role::Leader,
+    })
+}
+
+/// Construct a [`reqwest::Client`] suitable for use in a PPM [`Client`].
+pub fn default_http_client() -> Result<reqwest::Client, Error> {
+    Ok(reqwest::Client::builder()
+        .user_agent(CLIENT_USER_AGENT)
+        .build()?)
+}
+
 /// A PPM client.
 #[derive(Debug)]
-pub struct Client {
+pub struct Client<V, P, C> {
+    task_id: TaskId,
+    vdaf_client: V,
+    vdaf_public_parameter: P,
+    leader_url: Url,
+    clock: C,
     http_client: reqwest::Client,
     leader_report_sender: HpkeSender,
     helper_report_sender: HpkeSender,
 }
 
-impl Client {
+impl<V: VdafClient, C: Clock> Client<V, V::PublicParam, C> {
     pub fn new(
+        task_id: TaskId,
+        vdaf_client: V,
+        vdaf_public_parameter: V::PublicParam,
+        leader_url: Url,
+        clock: C,
         http_client: &reqwest::Client,
         leader_report_sender: HpkeSender,
         helper_report_sender: HpkeSender,
     ) -> Self {
         Self {
+            task_id,
+            vdaf_client,
+            vdaf_public_parameter,
+            leader_url,
+            clock,
             http_client: http_client.clone(),
             leader_report_sender,
             helper_report_sender,
         }
     }
 
-    /// Construct a [`reqwest::Client`] suitable for use in a PPM [`Client`].
-    //
-    // TODO: To be particularly useful, this function should return
-    // `Box<dyn JanusHttpClient>`, where `JanusHttpClient` is a trait we define
-    // that captures exactly what our client needs (e.g., GET). Then tests could
-    // provide an alternate implementation of it.
-    pub fn default_http_client() -> Result<reqwest::Client, Error> {
-        Ok(reqwest::Client::builder()
-            .user_agent(CLIENT_USER_AGENT)
-            .build()?)
-    }
+    /// Upload a [`crate::message::Report`] to the leader, per §4.3.2 of
+    /// draft-gpew-priv-ppm
+    pub async fn upload(&self, measurement: &V::Measurement) -> Result<(), Error> {
+        let input_shares = self
+            .vdaf_client
+            .shard(&self.vdaf_public_parameter, measurement)?;
 
-    pub async fn aggregator_hpke_sender(
-        http_client: &reqwest::Client,
-        task_id: TaskId,
-        aggregator_endpoint: Url,
-    ) -> Result<HpkeSender, Error> {
-        let hpke_config_response = http_client
-            .get(aggregator_endpoint.join("hpke_config")?)
+        // All supported VDAFs use two aggregators
+        assert_eq!(input_shares.len(), 2);
+
+        let nonce = Nonce {
+            time: Time::from_naive_date_time(self.clock.now()),
+            rand: rand::random(),
+        };
+
+        // No extensions supported yet
+        let extensions = vec![];
+
+        let associated_data = Report::associated_data(nonce, &extensions);
+
+        // Leader's share MUST be the first in the report. We assume/guess that
```

Comment:
(though fwiw, I couldn't find verbiage in the VDAF spec that for vectored data, data at the same index has to be sent to the same participant in all cases -- `run_vdaf` maintains this, but I'm not sure if it's normative)

Reply:
>In PPM, I think we need to care about keeping the order of the vectors generated by the underlying VDAF consistent, because the VDAF operations expect them to be consistent.

I agree with this, but we don't have text in either the PPM or VDAF specifications that declares this requirement.

Let's say that you call `Vdaf.measurement_to_shares` and it returns `[uncompressed_share, compressed_share]`. I don't think that there's anything in the text of VDAF or PPM currently that stops an implementation from sending `compressed_share` to the leader and `uncompressed_share` to the helper. This will fail, because the leader is expecting uncompressed shares and will reject the compressed shares for being the wrong size (same deal on the helper side). But the implementation that does this still complies with the specification.

To be clear, I don't think this is a problem in `janus` -- our implementation is doing the right thing. I want the protocol text to rule out this problem, hence the issue I filed on VDAF.